action/protocol/rewarding/reward.go

```diff
@@ -486,3 +493,32 @@ func (p *Protocol) assertLastBlockInEpoch(blkHeight uint64, epochNum uint64, rp
 	}
 	return nil
 }
+
+func (p *Protocol) recalculateCandidates(epochNum uint64, candidates state.CandidateList, kickoutlist *vote.Blacklist) state.CandidateList {
+	// resort based on blacklist
+	intensityRate := float64(uint32(100)-kickoutlist.IntensityRate) / float64(100)
+	candidatesMap := make(map[string]*state.Candidate)
+	updatedVotingPower := make(map[string]*big.Int)
+	for _, cand := range candidates {
+		candidatesMap[cand.Address] = cand
+		if _, ok := kickoutlist.BlacklistInfos[cand.Address]; ok {
+			// if it is an unqualified delegate, multiply the voting power with kick-out intensity rate
+			if intensityRate == float64(0) {
+				// in case of hard kickout, make sure not to be top 36 candidate
+				delete(updatedVotingPower, cand.Address)
+			}
+			votingPower := new(big.Float).SetInt(cand.Votes)
+			newVotingPower, _ := votingPower.Mul(votingPower, big.NewFloat(intensityRate)).Int(nil)
+			updatedVotingPower[cand.Address] = newVotingPower
+		} else {
+			updatedVotingPower[cand.Address] = cand.Votes
+		}
+	}
+	// sort again with updated voting power
+	sorted := util.Sort(updatedVotingPower, epochNum)
+	var verifiedCandidates state.CandidateList
+	for _, name := range sorted {
+		verifiedCandidates = append(verifiedCandidates, candidatesMap[name])
+	}
+	return verifiedCandidates
+}
```

Comment:
the code looks redundant. Let's figure out a better way to prevent the redundancy.

Reply:
To avoid code duplicate, refactor CandidatesByHeight/ReadBP~/ReadActiveBP~ to return filtered candidate list -- so reward, ioctl doesn't have to re-calculate